File: backend/lib/L2_data/models/goals/task.py
#  Copyright (c) 2022. Alexandr Moroz

from sqlalchemy import Column, ForeignKey, Integer
from sqlalchemy.orm import relationship

from ..base_model import BaseModel
from .smartable import Smartable


class Task(Smartable, BaseModel):
    parent_id = Column(Integer, ForeignKey("tasks.id", ondelete="CASCADE"))

    # parent = relationship("Task", remote_side="Task.id")
    tasks = relationship("Task", remote_side="Task.parent_id")

    goal_id = Column(Integer, ForeignKey("goals.id", ondelete="CASCADE"), nullable=False)

    status_id = Column(Integer, ForeignKey("taskstatuss.id", ondelete="SET NULL"))
    status = relationship("TaskStatus")

    priority_id = Column(Integer, ForeignKey("taskprioritys.id", ondelete="SET NULL"))
    priority = relationship("TaskPriority")

File: frontend/lib/L1_domain/entities/goals/task.dart
// Copyright (c) 2022. Alexandr Moroz

import 'smartable.dart';
import 'task_priority.dart';
import 'task_status.dart';

class Task extends Smartable {
  Task({
    required int id,
    required DateTime createdOn,
    required DateTime updatedOn,
    required String title,
    required String description,
    required DateTime? dueDate,
    required int? parentId,
    required int? trackerId,
    required bool closed,
    required List<Task> tasks,
    required this.status,
    required this.priority,
  }) : super(
          id: id,
          createdOn: createdOn,
          updatedOn: updatedOn,
          title: title,
          description: description,
          dueDate: dueDate,
          parentId: parentId,
          closed: closed,
          tasks: tasks,
          trackerId: trackerId,
        );

  final TaskStatus? status;
  final TaskPriority? priority;

  Task copy() => Task(
        id: id,
        parentId: parentId,
        createdOn: createdOn,
        updatedOn: updatedOn,
        title: title,
        description: description,
        dueDate: dueDate,
        status: status,
        priority: priority,
        tasks: tasks,
        closed: closed,
        trackerId: trackerId,
      );

  @override
  List<Task> get allTasks {
    final List<Task> res = [];
    for (Task t in tasks) {
      res.addAll(t.allTasks);
      res.add(t);
    }
    return res;
  }
}
